File: packages/protocol/src/schema.ts
export type Column = {
  name: string;
  type: string;
  nullable: boolean;
  default: string | null;
  pk: boolean;
  position: number;
};

/** One FOREIGN KEY constraint. Composite keys carry several columns, paired by index. */
export type ForeignKey = {
  /** The constraint name as the database reports it. */
  name: string;
  /** Columns on this table, in constraint order. */
  columns: string[];
  refSchema: string;
  refTable: string;
  /** Columns on the referenced table, paired with `columns` by index. */
  refColumns: string[];
  /** "NO ACTION" | "CASCADE" | "SET NULL" | "SET DEFAULT" | "RESTRICT" as the database spells it. */
  onDelete: string;
  onUpdate: string;
};

export type Table = {
  schema: string;
  name: string;
  kind: "table" | "view" | "materialized_view";
  columns: Column[];
  /** FOREIGN KEY constraints declared on this table; empty when none, and always empty for views. */
  foreignKeys: ForeignKey[];
  /** Planner estimate when cheap to get, else undefined. */
  rowEstimate?: number;
};

export type Schema = { name: string; tables: Table[] };

export type DatabaseSchema = {
  database: string;
  schemas: Schema[];
  fetchedAt: string;
};

File: apps/server/protocol/schema.go
package protocol

type Column struct {
	Name     string  `json:"name"`
	Type     string  `json:"type"`
	Nullable bool    `json:"nullable"`
	Default  *string `json:"default"`
	PK       bool    `json:"pk"`
	Position int     `json:"position"`
}

// Composite keys carry several columns, paired with RefColumns by index.
type ForeignKey struct {
	Name       string   `json:"name"`
	Columns    []string `json:"columns"`
	RefSchema  string   `json:"refSchema"`
	RefTable   string   `json:"refTable"`
	RefColumns []string `json:"refColumns"`
	OnDelete   string   `json:"onDelete"`
	OnUpdate   string   `json:"onUpdate"`
}

type TableKind string

const (
	KindTable            TableKind = "table"
	KindView             TableKind = "view"
	KindMaterializedView TableKind = "materialized_view"
)

type Table struct {
	Schema      string       `json:"schema"`
	Name        string       `json:"name"`
	Kind        TableKind    `json:"kind"`
	Columns     []Column     `json:"columns"`
	ForeignKeys []ForeignKey `json:"foreignKeys"`
	RowEstimate *int64       `json:"rowEstimate,omitempty"`
}

type Schema struct {
	Name   string  `json:"name"`
	Tables []Table `json:"tables"`
}

type DatabaseSchema struct {
	Database  string   `json:"database"`
	Schemas   []Schema `json:"schemas"`
	FetchedAt string   `json:"fetchedAt"`
}
